File: sdks/python/examples/shopping_cart/endpoint.py
from __future__ import annotations

from collections.abc import AsyncIterator

from ankka import Done, Endpoint, HttpProblem, delete, get, post, sse
from ankka.client import Calls, ComponentClient

from examples.shopping_cart.cart_rows import CartRow
from examples.shopping_cart.checkout_log import CheckoutRecord
from examples.shopping_cart.checkout_workflow import Checkout
from examples.shopping_cart.domain import LineItem, ShoppingCart


# docs:start endpoint
class ShoppingCartEndpoint(Endpoint):
    """The Scala sample's routes, exactly: /carts/{cartId}, /total, /items, /items/{productId}, /checkout."""

    prefix = "/carts"

    def __init__(self, client: ComponentClient) -> None:
        self.client = client

    def _cart(self, cart_id: str) -> Calls:
        return self.client.with_metadata(self.request.metadata).for_event_sourced_entity("shopping-cart", cart_id)

    @get("/{cartId}")
    async def get_cart(self, cartId: str) -> ShoppingCart:
        return await self._cart(cartId).call("get-cart").invoke(reply=ShoppingCart)

    @get("/{cartId}/total")
    async def total(self, cartId: str) -> int:
        return await self._cart(cartId).call("total-quantity").invoke(reply=int)

    @post("/{cartId}/items")
    async def add_item(self, cartId: str, item: LineItem) -> Done:
        return await self._cart(cartId).call("add-item").invoke(item, reply=Done)

    @delete("/{cartId}/items/{productId}")
    async def remove_item(self, cartId: str, productId: str) -> Done:
        return await self._cart(cartId).call("remove-item").invoke(productId, reply=Done)

    @post("/{cartId}/checkout")
    async def checkout(self, cartId: str) -> ShoppingCart:
        return await self._cart(cartId).call("checkout").invoke(reply=ShoppingCart)
    # docs:end endpoint

    # ── The view, the workflow and the notifier's log ──────────────────────

    @get("/awkward")
    def awkward(self) -> str:
        """A literal beside a parameter: the router must prefer it over ``/{cartId}``."""
        return "literal"

    # docs:start problem
    @get("/{cartId}/rows")
    async def row(self, cartId: str) -> CartRow:
        found = await self.client.views.get("cart-rows", cartId, CartRow)
        if found is None:
            raise HttpProblem(404, f"no row for cart '{cartId}'")
        return found  # type: ignore[no-any-return]
    # docs:end problem

    @get("/rows")
    async def rows(self) -> list[CartRow]:
        return await self.client.views.all("cart-rows", CartRow)

    # docs:start start-workflow
    @post("/{cartId}/checkouts")
    async def start_checkout(self, cartId: str, mode: str) -> Done:
        """``mode`` is the body: ``ok``, ``fail`` or ``pause``."""
        return await self.client.with_metadata(self.request.metadata).for_workflow("checkout", cartId).call("start").invoke(mode or "ok", reply=Done)
    # docs:end start-workflow

    @get("/{cartId}/checkouts")
    async def checkout_status(self, cartId: str) -> Checkout:
        return await self.client.with_metadata(self.request.metadata).for_workflow("checkout", cartId).call("status").invoke(reply=Checkout)

    # ── The assistant: a POST answers whole, a GET streams tokens as SSE ───

    # docs:start agent-routes
    @post("/ask/{session}")
    async def ask(self, session: str, question: str) -> str:
        return await self.client.with_metadata(self.request.metadata).for_agent("assistant", session).call("ask").invoke(question, reply=str)

    @sse("/chat/{session}")
    async def chat(self, session: str) -> AsyncIterator[str]:
        question = next((v for k, v in self.request.query if k == "q"), "")
        async for token in self.client.with_metadata(self.request.metadata).for_agent("assistant", session).call("chat").stream(question):
            yield token
    # docs:end agent-routes

    @get("/{cartId}/checkout-log")
    async def checkout_log(self, cartId: str) -> CheckoutRecord:
        return await self.client.with_metadata(self.request.metadata).for_key_value_entity("checkout-log", cartId).call("get").invoke(reply=CheckoutRecord)

File: sdks/typescript/examples/shopping-cart/endpoint.ts
import { Acl, Done, Endpoint, HttpProblem, del, get, post, s, sse } from "ankka"
import { LineItem, ShoppingCart } from "./domain.ts"
import { ShoppingCartEntity } from "./entity.ts"
import { CartRow, CartRows } from "./cartRows.ts"
import { Checkout, CheckoutWorkflow } from "./checkoutWorkflow.ts"
import { CheckoutLog, CheckoutRecord } from "./checkoutLog.ts"
import { CartAssistant } from "./assistant.ts"

// docs:start endpoint
/** The Scala sample's routes, exactly: /carts/{cartId}, /total, /items, /items/{productId}, /checkout. */
export class ShoppingCartEndpoint extends Endpoint {
  static readonly prefix = "/carts"
  static readonly acl = Acl.allowAll

  static readonly routes = {
    getCart: get("/{cartId}", ShoppingCart, (ep: ShoppingCartEndpoint, req) => ep.cart(req.params.cartId).call(ShoppingCartEntity.handlers.getCart).invoke()),
    total: get("/{cartId}/total", s.int, (ep: ShoppingCartEndpoint, req) => ep.cart(req.params.cartId).call(ShoppingCartEntity.handlers.totalQuantity).invoke()),
    addItem: post("/{cartId}/items", LineItem, Done, (ep: ShoppingCartEndpoint, req, item) => ep.cart(req.params.cartId).call(ShoppingCartEntity.handlers.addItem).invoke(item)),
    removeItem: del("/{cartId}/items/{productId}", Done, (ep: ShoppingCartEndpoint, req) =>
      ep.cart(req.params.cartId).call(ShoppingCartEntity.handlers.removeItem).invoke(req.params.productId),
    ),
    checkout: post("/{cartId}/checkout", ShoppingCart, (ep: ShoppingCartEndpoint, req) => ep.cart(req.params.cartId).call(ShoppingCartEntity.handlers.checkout).invoke()),
    // docs:end endpoint

    // A literal beside a parameter: the router must prefer it over /{cartId}.
    awkward: get("/awkward", s.string, () => "literal"),

    // ── The view, the workflow and the notifier's log ──
    // docs:start problem
    row: get("/{cartId}/rows", CartRow, async (ep: ShoppingCartEndpoint, req) => {
      const found = await ep.client.views.get(CartRows.componentId, req.params.cartId, CartRow)
      if (found === null) throw new HttpProblem(404, `no row for cart '${req.params.cartId}'`)
      return found
    }),
    // docs:end problem
    rows: get("/rows", s.list(CartRow), (ep: ShoppingCartEndpoint) => ep.client.views.all(CartRows.componentId, CartRow)),
    // docs:start start-workflow
    /** The body is the mode: `ok`, `fail` or `pause`. */
    startCheckout: post("/{cartId}/checkouts", s.string, Done, (ep: ShoppingCartEndpoint, req, mode) =>
      ep.client.of(CheckoutWorkflow, req.params.cartId).call(CheckoutWorkflow.handlers.start).invoke(mode || "ok"),
    ),
    // docs:end start-workflow
    checkoutStatus: get("/{cartId}/checkouts", Checkout, (ep: ShoppingCartEndpoint, req) =>
      ep.client.of(CheckoutWorkflow, req.params.cartId).call(CheckoutWorkflow.handlers.status).invoke(),
    ),
    checkoutLog: get("/{cartId}/checkout-log", CheckoutRecord, (ep: ShoppingCartEndpoint, req) =>
      ep.client.of(CheckoutLog, req.params.cartId).call(CheckoutLog.handlers.get).invoke(),
    ),

    // ── The assistant: a POST answers whole, a GET streams tokens as SSE ──
    // docs:start agent-routes
    ask: post("/ask/{session}", s.string, s.string, (ep: ShoppingCartEndpoint, req, question) =>
      ep.client.of(CartAssistant, req.params.session).call(CartAssistant.handlers.ask).invoke(question),
    ),
    chat: sse("/chat/{session}", (ep: ShoppingCartEndpoint, req) =>
      ep.client.of(CartAssistant, req.params.session).call(CartAssistant.handlers.chat).stream(req.query.get("q") ?? ""),
    ),
    // docs:end agent-routes
  }

  cart(cartId: string) {
    return this.client.of(ShoppingCartEntity, cartId)
  }
}
